File: src/main/kotlin/cl/ravenhill/pijamaparty/stinginthetail/Fibonacci.java
package cl.ravenhill.pijamaparty.stinginthetail;

public class Fibonacci {
  public static void main(String[] args) {
    Fibonacci fib = new Fibonacci();
    fib.printSequence(9);
  }

  public void printSequence(int n) {
    printSequenceAux(n, 0, 1);
  }

  public void printSequenceAux(int n, int a, int b) {
    System.out.println(a);
    if (n > 0) {
      // Tail recursion :0!
      printSequenceAux(n - 1, b, a + b);
    }
  }
}

File: src/main/kotlin/cl/ravenhill/pijamaparty/stinginthetail/Fibonacci.kt
package cl.ravenhill.pijamaparty.stinginthetail

fun main() {
  printFibonacci(9)
}

fun printFibonacci(n: Int) {
  fibAux(n, 0, 1)
}

tailrec fun fibAux(n: Int, a: Int, b: Int) {
  println(a)
  if (n > 0) {
    fibAux(n - 1, b, a + b)
  }
}
